libdnf/module/module_sack.cpp

```diff
@@ -145,6 +149,150 @@ BaseWeakPtr ModuleSack::get_base() const {
+            //     }rpm/package_query.hpp
+            // }
+            copy(std::begin(artifacts), std::end(artifacts), std::back_inserter(include_NEVRAs));
+        } else {
+            copy(std::begin(artifacts), std::end(artifacts), std::back_inserter(exclude_NEVRAs));
+        }
+    }
+
+    return std::make_tuple(
+        std::move(include_NEVRAs),
+        std::move(exclude_NEVRAs),
+        std::move(names),
+        std::move(src_names),
+        std::move(reldep_name_list));
+}
+
+void ModuleSack::Impl::module_filtering() {
+    auto [include_NEVRAs, exclude_NEVRAs, names, src_names, reldep_name_list] = collect_data_for_modular_filtering();
+
+    // Pakages from system, commandline, and hotfix repositories are not targets for modular filterring
+    libdnf::rpm::PackageQuery target_packages(base);
+
+    // TODO(replace) "@System", "@commandline" by defined variables like in dnf4
+    std::vector<std::string> keep_repo_ids = {"@System", "@commandline"};
+
+    libdnf::repo::RepoQuery hotfix_repos(base);
+    hotfix_repos.filter_enabled(true);
+    hotfix_repos.filter(
+        [](const libdnf::repo::RepoWeakPtr & repo) { return repo->get_config().module_hotfixes().get_value(); },
+        true,
+        libdnf::sack::QueryCmp::EQ);
+    for (auto & repo : hotfix_repos) {
+        keep_repo_ids.push_back(repo->get_id());
+    }
+
+    target_packages.filter_repo_id(keep_repo_ids, libdnf::sack::QueryCmp::NEQ);
+
+    libdnf::rpm::PackageQuery include_query(base);
+    libdnf::rpm::PackageQuery exclude_query(target_packages);
+    libdnf::rpm::PackageQuery exclude_provides_query(target_packages);
+    libdnf::rpm::PackageQuery exclude_names_query(target_packages);
+    libdnf::rpm::PackageQuery exclude_src_names_query(target_packages);
+
+    include_query.filter_nevra(include_NEVRAs);
+
+    // All packages from not active modules must be filtered out by modular filtering except packages from active
+    // modules
+    exclude_query.filter_nevra(exclude_NEVRAs);
+    exclude_query.difference(include_query);
+
+    // Exclude packages by their Provides. Provides are used to disable obsoletes. Remove included modular packages to
+    // not exclude modular packages from active modules
+    exclude_provides_query.filter_provides(reldep_name_list);
+    exclude_provides_query.difference(include_query);
+
+    // Search for source packages with same names as included source artifacts
+    exclude_src_names_query.filter_name(src_names);
+    exclude_src_names_query.filter_arch({"src", "nosrc"});
+
+    // Required to filtrate out source packages and packages with incompatible architectures.
+    exclude_names_query.filter_name(names);
+
+    // unifying with exclude_src_names_query will remove additional removal of included packages. Remove included
+    // modular packages to not exclude modular packages from active modules
+    exclude_names_query.update(exclude_src_names_query);
+    exclude_names_query.difference(include_query);
```

Comment:
I don't understand these comments. I thought the `names` are excluded, so that we exclude all packages that have the same name as the modular packages from active modules (but different from the modular packages from active modules obviously).
And why do we distinguish between `names` and `src_names` when we then exclude them both? Is it used somewhere else? If so, it deserves to be commented.

Reply:
I provided additional text above. Hopefully it will be more explanatory.